### PR TITLE
ARROW-10103: [Rust] Add contains kernel

### PR DESCRIPTION
**Pull request by nevi-me**
Checks if a list contains a value in either a primitive or string

Large lists are also supported

### CODE REVIEW

**Comment by nevi-me** [comment]
This is extracted from #6770, 

PTAL @alamb @jorgecarleitao @jhorstmann 

CC @mcassels @maxburke

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/comparison.rs:693]: Is it correct for us to reuse the offset, or should this be 0? My intuition says the latter, but I'm too tired to figure it out. Same applies above

File: rust/arrow/src/compute/util.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/util.rs:88]: Out of curiosity: is this true in general, or is for these (this and above) implementation, that uses this assumption?

File: rust/arrow/src/compute/util.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/util.rs:77]: tests missing, I think.

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/comparison.rs:600]: I think that we can short-circuit here (`result.append(true)?; break`) (and equivalent instances). In rust's notation, I think that we are using the [`Iterator::any`](https://doc.rust-lang.org/stable/rust-by-example/fn/closures/closure_examples/iter_any.html) with some null logic sprinkled on top.

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/comparison.rs:691]: I also think that this should be zero: we are building a new buffer (`result.finish()`) and thus there is no need to start from an offset. I would expect that a test with `left.offset() != 0` to not pass with this code, as we read `result`'s buffer from left's offset.

**Review by jorgecarleitao — COMMENTED** [review]
I went through this. Looks great. Thanks a lot, @nevi-me !

I think that `compare_option_bitmap` would benefit from some tests, and that there is a potential fix needed wrt to offsets.

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jhorstmann) [inline rust/arrow/src/compute/kernels/comparison.rs:620]: Offset should be 0 since the result is a new buffer that only contains the relevant data

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jhorstmann) [inline rust/arrow/src/compute/kernels/comparison.rs:645]: I think this is creating a buffer that is larger than needed since it is a bitmap storing 8 values per byte. Correct size would be `ceil(left.len(), 8)`

File: rust/arrow/src/compute/util.rs
Comment (jhorstmann) [reply rust/arrow/src/compute/util.rs:88]: I'm trying to solve this problem in #8262 

The issue is that `buffer.slice` and `buffer_bin_or/and` currently work with byte offsets, but for boolean arrays and bitmaps the offset can start in the middle of a byte.

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jhorstmann) [inline rust/arrow/src/compute/kernels/comparison.rs:598]: Here again we'd need to consider the offset of `left`, or use `left.is_valid` instead

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (jhorstmann) [inline rust/arrow/src/compute/kernels/comparison.rs:593]: The sql behaviour regarding nulls would be different, if that is intended it should probably be noted a bit more prominently. In Postgres:

```sql
SELECT
  'foo' = ANY(ARRAY['foo','bar']::text[]) AS non_null,
  'foo' = ANY(ARRAY[]::text[]) AS empty,
  'foo' = ANY(ARRAY[NULL]::text[]) AS null_value,
  'foo' = ANY(NULL::text[]) AS null_array
```

```
| non_null | empty | null_value | null_array |
| -------- | ----- | ---------- | ---------- |
| true     | false |            |            |
```
(empty cells in that table being `null` values)

Comment (nevi-me) [reply rust/arrow/src/compute/kernels/comparison.rs:593]: I think the test is 

```sql
SELECT
        'foo' = ANY(ARRAY['foo','bar']::text[]) AS non_null,
        'foo' = ANY(ARRAY[]::text[]) AS empty,
        NULL::text = ANY(ARRAY[NULL, 'foo']::text[]) AS null_value,
        'foo' = ANY(NULL::text[]) AS null_array
```

but nonetheless it still yields the same result.
I copied this from another PR mostly-verbatim, so I'm only scrutinising the code today myself.

I would prefer aligning with the SQL behaviour, that a null value can't be contained in an array if the array has nulls.
If users want to find out if an array has nulls, they can use the below or something better:

```rust
let array: ListArray<i32> = ...;

let ceil = bit_util::ceil(array.len(), 8);
let mut bools = MutableBuffer::new(ceil).
for i in -..array.len() {
  if array.is_valid(i) && array.value(i).null_count() > 0 {
    bools.set_bit(i);
  }
}

// create bool array from bools
```

File: rust/arrow/src/compute/util.rs
Comment (nevi-me) [reply rust/arrow/src/compute/util.rs:77]: Added, I ended up not using `compare_option_bitmap`, but it's a useful function to have

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/comparison.rs]: PTAL at how I've done it now. Instead of using a booelan builder, I set the bits directly on the buffer like we did with `take`

File: rust/arrow/src/compute/kernels/comparison.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/comparison.rs]: Changed, ta

**Comment by nevi-me** [comment]
@jhorstmann @jorgecarleitao I've had a look at this with a fresh head, and I've simplified the logic to:

Given a list array, return true if a non-null value exists in the array.

**Comment by jhorstmann** [comment]
> Given a list array, return true if a non-null value exists in the array.

Intuitively this makes sense, unfortunately the sql rules are slightly more complex and can also return null in several cases. I'm ok with leaving it like this for now, but we might need to change it later to align more closely with sql.

**Comment by velvia** [comment]
@nevi-me Hi there, I'm teammates with @mcassels and @maxburke at UrbanLogiq, and we'd like to check in on the status of this PR.   From looking over it, this is what I can gleam (the PR desc is just one sentence):

* This just has kernel primitives for contains for generic strings and primitives
* No specific support for dictionary-encoded string arrays?  I guess this could be built on the primitive kernel?
* Is a separate PR for the rest of the support, ie logical plan, etc. coming?   Would be interested in building on top of this to finish the PR as we would like to use it on a newer branch.
(we'd be happy to finish the rest of it or work on the follow on PR)

Would love to have an ETA....

Thanks!
-Evan

**Comment by nevi-me** [comment]
Hi @velvia, I pulled these changes from UL's fork while I was rebasing the changes from #6770. I only refactored the existing kernel functions there, but didn't expand any supported data types.

The behaviour does change from what you had on your fork (https://github.com/apache/arrow/pull/8280#discussion_r495433674), so perhaps you can comment on whether this is fine, or bear that in mind when updating your fork.

Regarding an ETA, the PR's still under review, so once it's approved, we'll be able to merge it in.

Yes, you can open JIRAs (https://issues.apache.org/jira/projects/ARROW) then work on top of this PR.

I've spent some time looking at the UL fork, and I think there might be changes there that the wider community would benefit from if you upstream them a bit more frequently. I understand that sometimes we might take longer than is ideal to complete PR reviews and merge them; but that's often a function of the sporadic availability of capacity from the Rust developers.

We've also been very reliant on 2 people on the Parquet side, but I've been spending more time on the codebase so I can start picking up PRs on Parquet.

**Comment by velvia** [comment]
@nevi-me thanks, will have a look

**Comment by maxburke** [comment]
> I've spent some time looking at the UL fork, and I think there might be changes there that the wider community would benefit from if you upstream them a bit more frequently. I understand that sometimes we might take longer than is ideal to complete PR reviews and merge them; but that's often a function of the sporadic availability of capacity from the Rust developers.

We would love to get our changes into the mainline Arrow, and that is our goal!, we've just been very constriained lately in how much people-power we can use to get the patches into a state our committers are happy with :slightly_smiling_face: 

Slowly but surely we'll have more patches coming!

**Review by jorgecarleitao — APPROVED** [review]
LGTM

**Review by andygrove — APPROVED** [review]
LGTM

**Pull request closed without merging**
